### PR TITLE
Climdex new implementation

### PR DESCRIPTION
**Pull request by stevekogo**

### CODE REVIEW

**Comment by stevekogo** [comment]
@rdstern  Have a look at the dialog and subdialog layout then give comments.

**Comment by stevekogo** [comment]
@africanmathsinitiative/developers Help review this.

**Comment by rdstern** [comment]
From Lily I was looking at the dialogue and sub-dialogues.

I really like the main dialogue.
I assume the indices now write back to the data frames.  If not, then this is a big and important step.  And (of course) the sheets will presumably either carry the other columns needed, or that would be done separately.

In the sub-dialogues we still have that long list of the 27 indices.  I would really like you to consider how you can support users to not just tick everything, but to work with these indices with a lot more understanding.  This is more challenging.  To me it would start with you understanding these indices and  then preparing a table (as we have for the different distributions in R), with the **properties** of each index.

So my starter includes the following:

a) Some are rainfall, others are temperature-based.
b) Some are only annual, while others could be either annual or monthly.
c) Some need a base period specified - others do not.
d) Some include a bootstrap, while others don't
e) Some are only tropical, others are temperate (e.g. snowfall), and others are both.

I am sure there are more, which will become apparent once you go through them carefully.

Once you have these properties in a table it should be possible to have a system in the dialogues that enables or disables particular indices, depending on what general properties of the data gave been specified. (e.g. if you specify monthly, then those that are only annual would be disabled.)

I wonder also whether you have 2 tabs, with one for rainfall and the other for temperatures.  I would be OK with an option being to choose either rainfall or temperatures and then those indices become available.  (Of course you may opt to allow "both" together.)  
If they are on separate tabs, then perhaps they could each be visible, without any need for scrolling.

It will take some iterations to get these working really simp0ly and smoothly.  It could be good to have the meeting in Cape Town as your deadline - though I hope that lots more climatic will be working by then.  But this would be powerful to show.

Roger

**Comment by stevekogo** [comment]
@africanmathsinitiative/developers This is partially ready and I welcome reviews mainly on the R-code as I fix the code in the dialog.

**Comment by stevekogo** [comment]
@rdstern I have implemented writing indices to two data frames, one for yearly and one for mothly with the links implemented. I will now go ahead  and separate the indices into two categories  namely:- "annual" and "annual_monthly". This is just the start but we may need to have categories like "temperature_annual", "temperature_annual_monthly", "rainfall_annual", "rainfall_annual_monthly", "rainfall_temperature_annual" and "rainfall_temperature_annual_monthly". My main worry on the later case is having a large subdialog to accommodate six groupboxes.

**Comment by rdstern** [comment]
That sounds like very good progress.  I wonder if the main split in the indices should be between temperature and rainfall?  Then there is a checkbox (or two radio buttons) for **Annual or Monthly**.  If you opt for **Monthly**, then some of the indices remain visible, but are **disabled**, i.e. those that are only annual.  

Did you also like the idea of adding an R-Help button to show the description of the index from the R-manual?

**Comment by stevekogo** [comment]
@africanmathsinitiative/developers This is ready for review. There is a lot to look at from design of dialog and subdialog. @dannyparsons  may have a look at the r-methods.

**Comment by stevekogo** [comment]
@rdstern I have made a number of changes in the way the indices are clustered. This seems quite sensible in the user point of view but the indices doesn't seem to balance in the dialog since temperature indices are more. Here are screen shots the indices.
precipitation
temp

File: instat/static/InstatObject/R/instat_object_R6.R [hunk 1]
```diff
@@ -1108,4 +1108,80 @@ instat_object$set("public", "get_link_names", function(data_name, include_overal
...
+    my_data = cbind(yy1, split_data)
+    names(my_data) = c("Year_month", names(indices), "Year", "Month")
+    my_data <- my_data[c(1,3,4,2)]
+    my_data$Year = as.integer(as.character(my_data$Year))
+    my_data$Month = as.integer(as.character(my_data$Month))
+    year_col = self$get_climatic_column_name(data_name, year_label)
+    month_col = self$get_climatic_column_name(data_name, month_label)
+    key_list = list("Year", "Month")
+    names(key_list) = c(as.name(year_col), as.name(month_col))
+    if(self$get_linked_to_data_name(data_name, c(Year="Year", Month = "Month"))==""){
```
Comment (stevekogo) [inline instat/static/InstatObject/R/instat_object_R6.R]: key_list will be fine in this case.

File: instat/static/InstatObject/R/instat_object_R6.R [hunk 2]
```diff
@@ -1108,4 +1108,80 @@ instat_object$set("public", "get_link_names", function(data_name, include_overal
...
+      warning("monthly_data is missing, it will be created.")
+      data_list = list(my_data)
+      names(data_list) = paste(data_name, "monthly", sep = "_")
+      self$import_data(data_tables = data_list)
+      self$add_key(paste(data_name, "monthly", sep = "_"), c("Year", "Month"))
+      self$add_link(from_data_frame = data_name, to_data_frame = paste(data_name, "monthly", sep = "_"), link_pairs = key_list, type = keyed_link_label)
+    }
+    else{
+      self$merge_data(by=c("Year_month","Year", "Month"), data_name = self$get_linked_to_data_name(data_name, key_list), new_data = my_data )
```
Comment (stevekogo) [inline instat/static/InstatObject/R/instat_object_R6.R]: I am taking out the **by** argument then merge_data will find the columns from data to use for merging.

File: instat/static/InstatObject/R/instat_object_R6.R [hunk 3]
```diff
@@ -1108,4 +1108,80 @@ instat_object$set("public", "get_link_names", function(data_name, include_overal
...
+      self$merge_data(by=c("Year"), data_name=self$get_linked_to_data_name(data_name, key_list), new_data = my_data )
+    }
+  }
+}
+)
+
+ instat_object$set("public", "get_combined_metadata_value", function(data_name, property) {
+   if(!self$is_metadata(data_name, property))stop(paste(property, " is not a metadata column."))
+   #new_data = self$get_combined_metadata()
+   return(self$get_combined_metadata()[data_name, property])
```
Comment (stevekogo) [inline instat/static/InstatObject/R/instat_object_R6.R]: I tried using this but it doesn't seem to return anything. Maybe you can try it or provide an example on its usage.

**Comment by stevekogo** [comment]
@dannyparsons I have implemented most of the suggestions and the dialogs functionaality has greatly improved. However, I do not agree on your comment 
**"Finally, a small point, I noticed that the indices on the sub dialog for rain were disabled when I didn't have something in the rain receiver on the main dialog. I don't think we should do this. As a general principle is we don't force people to fill in dialog/sub dialog in a particular order. You might then have to deal carefully with cases of choosing indices without having the correct variables to do them,**"
since it seems something sensible to the user to have this property. @rdstern may give his opinion regarding this then I can go ahead to make any necessary changes.

**Comment by stevekogo** [comment]
I have implemented help on the indices through a new command button. This works fine but I still doubt if help should go the help in subdialog button.
As with the functionality, @rdstern may have rigorous testing on this and also make suggestions on how the layout can be improved.

**Comment by rdstern** [comment]
Sounds good and happy to test.  One small thing is that your earlier version had the number for each index.  I think it may be very useful for user's cross-referencing if we still have those "official" numbers in your new layout.

**Comment by stevekogo** [comment]
I took out the numbers since the number of indices in the group boxes are
few. Check the new implementation then advice if this feature is still
sensible to have.

**Comment by rdstern** [comment]
It is just when I look at [http://etccdi.pacificclimate.org/list_27_indices.shtml](url) the indices are numbered, so this would assist users to know exactly which one is being computed.  Perhaps the numbers could be in brackets at the end, rather than at the start if you wish.

File: instat/static/InstatObject/R/instat_object_R6.R (hunk 2 above)
Comment (dannyparsons) [inline instat/static/InstatObject/R/instat_object_R6.R]: If you remove the by, then the merge will only work when you have columns of the same name. So we need to specify the by still, as a named vector with the names in each data frame.

**Comment by stevekogo** [comment]
I have made a number of edits on this as per @rdstern suggestions and this is almost ready. @africanmathsinitiative/developers should find time to review this since it is urgent.

File: instat/static/InstatObject/R/instat_object_R6.R (hunk 3 above)
Comment (dannyparsons) [inline instat/static/InstatObject/R/instat_object_R6.R]: This worked fine for me: `InstatDataObject$get_data_objects("survey")$get_metadata(is_calculated_label)`

**Comment by dannyparsons** [comment]
Working much better and looks better too. I still get the following bugs:

- Doing column summaries first and then climdex doesn't work. The error is in merging climdex into existing data frame. The column types for year don't match.

- When my year column was called `Y`, and I did column summaries and then climdex it produced different data frames. So I end up with 3 data frames instead of 1, the original, the column summaries and the climdex summaries. I should only have two. Also, strangely in this case, although I did 5 summaries in climdex, they kept writing over the second column, so I always ended up with only one summary. I think this is all related to `Year` being hard coded in places in the R code. This is where testing with different data sets and different situations should be something you think about before submitting to find these issues.

- If I don't open the sub dialog, I don't get any indices. Even though there is one checked by default. This is a bug, but I think we should also discuss what are the defaults. I wouldn't mind if nothing was selected by default so that you think about what summaries you want to do. 
 
- I shouldn't be able to click ok when no indices have been selected.

On the design (might not be so important to get perfect now):

- I like the numbers at the end of the summaries. I think it would look better if there was a space before the number so `...Index [19]` instead of `...Index[19]` looks so much better.
 
- I think the alignment of all the checkboxes is the same i.e. all have same x position but the vertical spacing between them is inconsistent. Some are squashed close to each other and some are further apart. We might not want the dialog to be very large but it looks a bit cluttered at the moment.
 
- The Settings tab is more messy. Probably there need some thought about that layout and how to group things. But a quick improvement would be to get alignment and sizes of things to be consistent.

**Comment by stevekogo** [comment]
@dannyparsons Thanks for the comments. Regarding the issue of doing column summaries followed by climdex summaries, I noticed that once you convert the year column to factor, it loses it climatic property hence it cannot be accessed by the climdex indices.

**Comment by dannyparsons** [comment]
Ok I hadn't noticed that, good point too! We should fix that. 
However, I did the define climatic after converting Year to factor and it still doesn't work.

**Comment by stevekogo** [comment]
@dannyparsons I think this is in a fair shape to go in for use at the conference and further testing. Any additional concerns can be sorted out later.

**Comment by dannyparsons** [comment]
Yes I agree this will be good for demonstration purposes. Although there are still improvements needed from my comments above I believe. If they're not going to get done before today then they should be done very soon after then.

**Pull request merged**